File: Difficulty: Easy/Bird and maximum fruit gathering/bird-and-maximum-fruit-gathering.js
//{ Driver Code Starts
function main() {
    const t = parseInt(readLine());
    for (let i = 0; i < t; i++) {
        const totalTime = parseInt(readLine());
        const arr = readLine().split(' ').map(Number);
        const solution = new Solution();
        console.log(solution.maxFruits(arr, totalTime));
        console.log("~");
    }
}

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => { inputString += inputStdin; });

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(string => string.trim());
    main();
});

function readLine() { return inputString[currentLine++]; }

// } Driver Code Ends

class Solution {
    maxFruits(arr, totalTime) {
        let n = arr.length;

        // If totalTime is greater than or equal to the number of trees, return the sum of all fruits
        if (totalTime >= n) {
            return arr.reduce((sum, value) => sum + value, 0);
        }

        // Initialize the sliding window sum for the first 'totalTime' trees
        let currSum = arr.slice(0, totalTime).reduce((sum, value) => sum + value, 0);
        let maxFruits = currSum;

        // Use sliding window to find the maximum sum
        for (let i = totalTime; i < n; i++) {
            currSum += arr[i] - arr[i - totalTime];
            maxFruits = Math.max(maxFruits, currSum);
        }

        return maxFruits;
    }
}

File: Difficulty: Easy/Bird and maximum fruit gathering/bird-and-maximum-fruit-gathering.java
//{ Driver Code Starts
// Initial Template for Java

// Initial Template for Java

import java.io.*;
import java.util.*;


// } Driver Code Ends

class Solution {
    long maxFruits(long arr[], int m) {
        // Get the length of the array
        int n = arr.length;
        
        // If the number of elements in the array is less than or equal to m, 
        // simply return the sum of all elements in the array
        if (n <= m) {
            long total = 0;
            for (long fruit : arr) {
                total += fruit;
            }
            return total;
        }

        // Initialize the variables to track the maximum sum and the sum of the current window
        long maxSum = 0, windowSum = 0;

        // Calculate the sum of the first 'm' elements (the initial window)
        for (int i = 0; i < m; i++) {
            windowSum += arr[i];
        }
        
        // Set the initial max sum to the sum of the first 'm' elements
        maxSum = windowSum;

        // Use a sliding window to calculate the maximum sum of any m consecutive elements
        for (int i = m; i < (n + m) - 1; i++) {
            // Update the window sum by adding the next element and removing the element that is sliding out of the window
            windowSum += arr[i % n] - arr[(i - m) % n];
            
            // Update the max sum if the current window sum is greater
            if (windowSum > maxSum) {
                maxSum = windowSum;
            }
        }

        // Return the maximum sum found
        return maxSum;
    }
}



//{ Driver Code Starts.

// Driver class
class Array {

    // Driver code
    public static void main(String[] args) throws IOException {
        // Taking input using buffered reader
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int testcases = Integer.parseInt(br.readLine());

        while (testcases-- > 0) {

            String line1 = br.readLine();
            String[] a1 = line1.trim().split("\\s+");
            int n = a1.length;
            long a[] = new long[n];
            for (int i = 0; i < n; i++) {
                a[i] = Long.parseLong(a1[i]);
            }
            int m = Integer.parseInt(br.readLine());
            Solution ob = new Solution();

            long ans = ob.maxFruits(a, m);
            System.out.println(ans);
            System.out.println("~");
        }
    }
}

// } Driver Code Ends